### ISSUE TITLE
CH6-Margin Maximization

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learning-MIT#13]
### margin 最靠近直線的資料點到直線的距離

Consider the following data, and two potential separators (red and blue).
```PY
data = np.array([[1, 2, 1, 2, 10, 10.3, 10.5, 10.7],
                 [1, 1, 2, 2,  2,  2,  2, 2]])
labels = np.array([[-1, -1, 1, 1, 1, 1, 1, 1]])
blue_th = np.array([[0, 1]]).T
blue_th0 = -1.5
red_th = np.array([[1, 0]]).T
red_th0 = -2.5
```

The situation is illustrated in the figure below.
![image](https://user-images.githubusercontent.com/34474924/233310141-1d39f23b-062e-429c-9f8c-1910f288b802.png)

What are the values of each score  (S sum ​ ,S min ​ ,S max ​ ) on the red separator?
Solution: [31.5, -1.5, 8.2]

Which of these separators maximizes  S sum ​ ?
Solution: red

Which separator maximizes S min ​ ? 這才是主要要看的!
Solution: blue

1F) Which score function should we prefer if our goal is to find a separator that generalizes better to new data?
Solution: S_min

S min ​is the score function that picks the blue separator, which is preferred in this case because it does not make any mistakes on the training data while having the same complexity as the red separator.

**Comment by Pin-Jiun** [comment]
### What a loss
Recall that the margin of a given point is defined as

![image](https://user-images.githubusercontent.com/34474924/233533369-843531ba-f40f-42d0-8a8c-dd643bbb217d.png)



Based on the previous part, we've decided to try to find a linear separator that maximizes the minimum margin (the distance between the separator and the points that come closest to it.)

We define the margin of a data set (X,Y), with respect to a separator as
![image](https://user-images.githubusercontent.com/34474924/233314308-05d159a3-019f-4e20-88ab-ad2eb17ad1ae.png)

an approach to this problem is to specify a value γ ref ​ for the margin of the data set, and then seek to find a linear separator that maximizes γ ref ​.

![image](https://user-images.githubusercontent.com/34474924/233314411-f207f904-4085-40c8-a57a-74dcc61ca4aa.png)




Suppose for our data set we find that the maximum γ ref ​ across all linear separators is 0.
Is our data linearly separable?
Solution: No
Intuitively, a maximum γ ref ​ of 0 says that our best separator is distance 0 from at least two data points having opposite labels. (Otherwise, we could move the separator off both of these points and achieve positive margin.) Thus, these data points, and our data set as a whole, cannot be separated by a hyperplane.

2C) For this subproblem, assume that γ ref ​ >0 (i.e., the data is linearly separable). Note that in this case, the Perceptron algorithm is guaranteed to find a separator that correctly classifies all of the data points.
What is the largest minimum margin guaranteed by running the Perceptron algorithm on a data set that has a maximum margin equal to γ ref ​ >0?

![image](https://user-images.githubusercontent.com/34474924/233529837-e179b643-c7c6-40e3-bc07-109e21d49474.png)
Solution: some ϵ where ϵ>0
Explanation: Since the maximum minimum margin is equal to some  γ ref ​ >0, we know that the data is linearly separable. Thus, Perceptron is guaranteed to correctly classify all data points, implying the resulting margin will be >0. We can also upper-bound the resulting margin by  γ ref ​ , since this is given as the largest possible margin for a linear separator on this dataset. However, Perceptron does not necessarily find the separator with the largest margin. Beyond whether the points are classified correctly, Perceptron pays no attention to the margin. That is, Perceptron could converge on a separator that is arbitrarily close to any of the points as long as the points are classified correctly.

____________

A typical form of the optimization problem is to minimize an objective that has the form
![image](https://user-images.githubusercontent.com/34474924/233316285-85424861-a390-4ab9-84f6-6c1369c0851f.png)

We first consider the objective of finding a maximum-margin separator using the format above, using the so-called "zero-infinity" loss

![image](https://user-images.githubusercontent.com/34474924/233316393-73649341-9488-44fc-b765-c65b7a54908b.png)
![image](https://user-images.githubusercontent.com/34474924/233316440-38bc39ce-b930-46a3-811f-ff9887d85314.png)

For a linearly separable data set, positive λ, and positive R given nonzero θ, what is true about the minimal value of  J 0,∞ ​
Which of the following is true:
![image](https://user-images.githubusercontent.com/34474924/233317942-ea4bd2bd-8486-4e06-936f-5b728d13da9b.png)

Solution: It is always finite and positive
Explanation:
If the data is linearly separable, there exists a separator that classifies perfectly, so the only term in the objective function is the regularization term, which will must be positive and finite, given finite and positive R.

For a non linearly separable data set, and positive λ and  γ ref ​ , what is true about the minimal value of J 0,∞ ​ :
Solution: It is infinite

**Comment by Pin-Jiun** [comment]
### Hinge Loss

in real data sets it is relatively rare that the data is linearly separable, so our algorithm should be able to handle this case also and still work toward an optimal, though imperfect, linear separator.

Instead of using ( 0 , ∞ ) loss, we should design a loss function that will let us "relax" the constraint that all of the points have margin bigger than  γ ref ​ , while still encouraging large margins.

The `hinge loss` is one such more relaxed loss function; we will define it in a way that makes a connection to the problem we are facing:

![image](https://user-images.githubusercontent.com/34474924/233319649-0746a21f-c154-4bb0-b915-3bf155d91f8c.png)

When the margin of the point is greater than or equal to  γ ref ​ , we are happy and the loss is 0; while when the margin is less than γ ref ​ , we have a positive loss that increases the further away the margin is from γ ref ​ .

3A) Given this definition, if  γ ref ​ is positive what can we say about L ℎ , no matter what finite values θ and θ0 ​ take on?
Solution: It is always >= 0


Here is a separator and three points. The dotted lines represent the margins determined by γ ref ​ .
```PY
data = np.array([[1.1, 1, 4],[3.1, 1, 2]])
labels = np.array([[1, -1, -1]])
th = np.array([[1, 1]]).T
th0 = -4
```

![image](https://user-images.githubusercontent.com/34474924/233532392-16ec586d-1f2e-425a-835f-2d4279008d32.png)


What is Lh for each point, where  γ ref ​ = √2 ​ /2? Enter the values in the same order as the respective points are listed in data. You can do this computationally or by hand.
Solution: [0.8, 0.0, 3.0]

![image](https://user-images.githubusercontent.com/34474924/233534792-9b32487b-0fed-416e-831b-c8d231e44cdf.png)

**Comment by Pin-Jiun** [comment]
### 4)  SVM (support vector machine)支援向量機 objective
Putting hinge loss and regularization together, we can look at regularized average hinge loss:

![image](https://user-images.githubusercontent.com/34474924/233548811-1d2efb75-af6b-4b0a-8841-226e153d6140.png)
![image](https://user-images.githubusercontent.com/34474924/233548970-0eb8dbd6-2e57-4196-84fe-9aa8d5f1d0a2.png)

4A) If the data is linearly separable and we use the SVM objective, if we now let λ=0 and find the minimizing values of θ,θ0 ​ , what will happen?
Solution: The minimal objective value will be 0
![image](https://user-images.githubusercontent.com/34474924/233550236-847ec72f-ff4d-4d85-b933-8cc8a4596579.png)

__________________
### 資料點相同, λ不同對separators的影響
Hint: 以下的圖都是經過最佳化所得到的結果, `J(θ,θ0​)`是相同的
4B) Consider the following plots of separators. They are for λ values of 0 and 0.001. Match each λ to a plot.
![image](https://user-images.githubusercontent.com/34474924/233550526-5ca00bdc-0124-475a-b93f-ab0253275e0b.png)
∥θ∥a = ((-0.0737901)**2+2.40847205**2)**0.5 = 2.409602165190182, 同理∥θ∥b =2.5677

`∥θ∥ : a<b` , 因為γref​=1/∥θ∥, 故 γref​ a>b
但是γ差異不明顯, 可以視為Lh是差不多的(都趨近為0)

因為J(θ,θ0​)是相同的, λ∥θ∥**2 也要相同, 故λa >λb

[lambdaA, lambdaB] Solution: [0.001, 0.0]
![image](https://user-images.githubusercontent.com/34474924/233550604-5a5a5818-910b-4d01-a14e-108f0e19ec8a.png)


4C) Consider the following three plots of separators. They are for λ values of 0, 0.001, and 0.03. Match to the plot.
![image](https://user-images.githubusercontent.com/34474924/233551857-c2927e5f-a34a-4cf1-9825-bedab1b08fea.png)

[lambdaA, lambdaB, lambdaC] Solution: [0.03, 0.0, 0.001]
λ越大表示越不會overfitting, 所以有可能造成training的時候misclassification的情況, 故A一定是0.03
剩下B跟C, 判斷方式和4B相同

兩者都完美的分割資料集, 可以視為J(θ,θ0​)是相同的
因為`∥θ∥ : B>C` , 所以　`λB<λC`

![image](https://user-images.githubusercontent.com/34474924/233552319-b69f07d6-b08a-4f9f-997b-badec87d52ee.png)

**Comment by Pin-Jiun** [comment]
### Linear Support Vector Machines
The training objective for the Support Vector Machine (with slack) can be seen as optimizing a balance between the average hinge loss over the examples and a regularization term that tries to keep θ small (or equivalently, increase the margin).
This balance is set by the regularization parameter `λ`. Here we only consider the case without the offset parameter θ0 ​ (setting it to zero) and rewrite the training objective as an average so that it is given by
![image](https://user-images.githubusercontent.com/34474924/233574617-3719f9c1-ba29-4cbf-b1e0-444cc9df63f7.png)
![image](https://user-images.githubusercontent.com/34474924/233575036-131c30e7-f4f9-44e4-8c19-9ed39f03147f.png)

In this problem we will optimize the training objective using a single training example, so that we can gain a better understanding of how the regularization parameter, λ, affects the result. To this end, we refer to the single training example as the feature vector and label pair, (x,y). We will then try to find a θ that minimizes

![image](https://user-images.githubusercontent.com/34474924/233575385-0f525753-8377-4d04-bcd3-c405bac77749.png)
![image](https://user-images.githubusercontent.com/34474924/233575440-2eace354-5b0f-4221-b0df-71d6a90ca937.png)
